src/snabbdom.ts

```diff
@@ -252,45 +275,45 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
 
   function patchVnode(oldVnode: VNode, vnode: VNode, insertedVnodeQueue: VNodeQueue) {
     let i: any, hook: any;
-    if (isDef(i = vnode.data) && isDef(hook = i.hook) && isDef(i = hook.prepatch)) {
+    if (isDef(i = vnode.data)) hook = i.hook;
+    if (isDef(hook) && isDef(i = hook.prepatch)) {
       i(oldVnode, vnode);
     }
     const elm = vnode.elm = (oldVnode.elm as Node);
-    let oldCh = oldVnode.children;
-    let ch = vnode.children;
     if (oldVnode === vnode) return;
     if (vnode.data !== undefined) {
       for (i = 0; i < cbs.update.length; ++i) cbs.update[i](oldVnode, vnode);
-      i = vnode.data.hook;
-      if (isDef(i) && isDef(i = i.update)) i(oldVnode, vnode);
+      if (isDef(hook) && isDef(i = hook.update)) i(oldVnode, vnode);
     }
+    let oldCh = oldVnode.children;
```

Comment:
these is another important change. `children` for old and new vnodes are extracted from the vnode after `update` patch (same for create), which means that a module can attempt to normalize the children collection. This is needed to emulate slotting of vnodes, etc. to match the shadow DOM slotting process.

It was doable before but requires developers to mutate the original children array, which wasn't very elegant, and an anti-pattern IMO.

Reply:
> This is needed to emulate slotting of vnodes, etc. to match the shadow DOM slotting process.

I implemented something like that ( prototype at https://codepen.io/blikblum/pen/BJrqyM ,  package coming soon)

I  implemented as a high order function (withSlot)

Could not figure how to implement as a module. Curious if/how these changes will allow a different implementation

Reply:
oh, in the past, we have done this by doing this in a module:

```js
   const slottedChildren = slotMyChildren(vnode.children, vnode.data.withSlot);
   vnode.children.length = 0;
   vnode.children.push.apply(vnode.children, slottedChildren);
```

which is nasty because you're mutating the original `children` provided in the vnode, which could have undesired side effects if memoized at any level in your app.

With the new code in this branch, you could improve that by doing:

```js
   vnode.children = slotMyChildren(vnode.children, vnode.data.withSlot);
```

sill you have the memoization process, but only if you cache the result of calling `h`.

Reply:
> oh, in the past, we have done this by doing this in a module

What hook did you used? `create`?

According to docs `create` is called after DOM element creation so it could lead to issues. That's why i did not tried it.

But reading your comment in this PR, i realized that is called before element creation so is possible to use `create` for such task

> In the previous iteration, the data.hook.create happens after the children is processed, while the > module.create happens before. In this refactor, they are unified, and are happening at the same > time. I was confused on the "why", could not really explain it.

Maybe to allow mutating vnode before element creation, as above?

Reply:
oh, no, this operation is not tied to `create` only, it must be done in `update` as well because the children collection might change for an already created vnode/elm, in which case you will have apply the slotting mechanism again.